tests/unit/s2n_server_key_share_extension_test.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include <stdint.h>
+#include <s2n.h>
+
+#include "tls/extensions/s2n_server_key_share.h"
+#include "tls/extensions/s2n_client_key_share.h"
+
+#include "testlib/s2n_testlib.h"
+#include "stuffer/s2n_stuffer.h"
+#include "utils/s2n_safety.h"
+
+static int s2n_public_ecc_keys_are_equal(struct s2n_ecc_params *param_1, struct s2n_ecc_params *param_2);
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test s2n_extensions_server_key_share_send_size */
+    {
+        struct s2n_connection *conn;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_EQUAL(0, s2n_extensions_server_key_share_send_size(conn));
+
+        conn->secure.server_ecc_params.negotiated_curve = &s2n_ecc_supported_curves[0];
+        EXPECT_EQUAL(( 32 * 2 ) + 1 + 8, s2n_extensions_server_key_share_send_size(conn));
+
+        conn->secure.server_ecc_params.negotiated_curve = &s2n_ecc_supported_curves[1];
+        EXPECT_EQUAL(( 48 * 2 ) + 1 + 8, s2n_extensions_server_key_share_send_size(conn));
+
+        conn->secure.server_ecc_params.negotiated_curve = NULL;
+        EXPECT_EQUAL(0, s2n_extensions_server_key_share_send_size(conn));
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test s2n_extensions_server_key_share_send */
+    {
+        struct s2n_connection *conn;
+
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        struct s2n_stuffer* extension_stuffer = &conn->handshake.io;
```

Comment:
Why reuse the same stuffer for all your writes? If you don't, you won't need your new test macros and will be able to use more standard stuffer functions to examine your output afterwards.

Reply:
reusing the stuffer and the new test macros are independent of each other. eg. EXPECT_STUFFER_BYTES_WRITTEN() can be redefined as

```
EXPECT_SUCCESS(s2n_stuffer_skip_read(stuffer, expected_size));
EXPECT_EQUAL(s2n_stuffer_data_available(stuffer), 0);
```

`STUFFER_READ_U16_EXPECT_EQUAL` is also defined with standard stuffer functions.

The question is whether there's an opinion on
a) defining new macros within the test or the shared lib
b) be a little more verbose and avoid macros

As for reusing stuffer, I think it can be done both ways. Reusing results in less code as the handshake stuffer gets freed together with the connection.